kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs

```diff
@@ -804,13 +842,181 @@ impl<'tcx> GotocCtx<'tcx> {
-        let output = self.codegen_ty(substs.as_generator().return_ty());
-        Type::code_with_unnamed_parameters(tys, output)
+    /// Translate a generator type similarly to an enum with a variant for each suspend point.
+    ///
+    /// Consider the following generator:
+    /// ```
+    /// || {
+    ///     let a = true;
+    ///     let b = &a;
+    ///     yield;
+    ///     assert_eq!(b as *const _, &a as *const _);
+    ///     yield;
+    /// };
+    /// ```
+    ///
+    /// Rustc compiles this to something similar to the following enum (but there are differences, see below!):
+    ///
+    /// ```ignore
+    /// enum GeneratorEnum {
+    ///     Unresumed,                        // initial state of the generator
+    ///     Returned,                         // generator has returned
+    ///     Panicked,                         // generator has panicked
+    ///     Suspend0 { b: &bool, a: bool },   // state after suspending (`yield`ing) for the first time
+    ///     Suspend1,                         // state after suspending (`yield`ing) for the second time
+    /// }
+    /// ```
+    ///
+    /// However, its layout may differ from normal Rust enums in the following ways:
+    /// * Contrary to enums, the discriminant may not be at offset 0.
+    /// * Contrary to enums, there may be other fields than the discriminant "at the top level" (outside the variants).
+    ///
+    /// This means that we CANNOT use the enum translation, which would be roughly as follows:
+    ///
+    /// ```ignore
+    /// struct GeneratorEnum {
+    ///     int case; // discriminant
+    ///     union GeneratorEnum-union cases; // variant
+    /// }
+    ///
+    /// union GeneratorEnum-union {
+    ///     struct Unresumed variant0;
+    ///     struct Returned variant1;
+    ///     // ...
+    /// }
+    /// ```
+    ///
+    /// Instead, we use the following translation:
+    ///
+    /// ```ignore
+    /// union GeneratorEnum {
+    ///     struct DirectFields direct_fields;
+    ///     struct Unresumed generator_variant_Unresumed;
+    ///     struct Returned generator_variant_Returned;
+    ///     // ...
+    /// }
+    ///
+    /// struct DirectFields {
+    ///     // padding (for bool *b in Suspend0 below)
+    ///     char case;
+    ///     // padding (for bool a in Suspend0 below)
+    /// }
+    ///
+    /// struct Unresumed {
+    ///     // padding (this variant has no fields)
+    /// }
+    ///
+    /// // ...
+    ///
+    /// struct Suspend0 {
+    ///     bool *generator_field_0; // variable b in the generator code above
+    ///     // padding (for char case in DirectFields)
+    ///     bool generator_field_1; // variable a in the generator code above
+    /// }
+    /// ```
+    ///
+    /// Of course, if the generator has any other top-level/direct fields, they'd be included in the `DirectFields` struct as well.
+    fn codegen_ty_generator(&mut self, ty: Ty<'tcx>) -> Type {
```

Comment:
Please add a debug log in this function.

Reply:
Done.